docs/go-openssl-compat.md

```diff
@@ -0,0 +1,87 @@
+
+This is because OpenSSL don't expect that recipe to be used for AES-GCM TLS encryption, but it provided instead a control parameter, EVP_CTRL_AEAD_TLS1_AAD, that when applied to an encryption context it makes it work in TLS-mode and completely change the encryption workflow (more details [here](https://beta.openssl.org/docs/manmaster/man3/EVP_CipherInit_ex.html#tlsivfixed-OSSL_CIPHER_PARAM_AEAD_TLS1_IV_FIXED-octet-string)).
+
+The main difference is that the explicit IV field is no longer managed by the caller but by the OpenSSL encryption context, incrementing it on every Seal operation and applying some additional checks. This means that OpenSSL won't honor the IV constructed by Go, and therefore it will make Go send corrupted packaged to the wire.
+
+#### Assumptions
+
+1. Go constructs IVs that are FIPS 140-2 compliant. BoringCrypto also works under this assumption.
+
+#### Options
+
+1. Use OpenSSL AES-GCM in TLS mode, which means IV construction is secured inside OpenSSL context but Go expectations are not meet, therefore it requires the following changes to Go TLS stack:
+  - In `go-crypto-openssl`, change `NewGCMTLS` to accept the 4 bytes fixed IV field.
+  - In `crypto/tls`:
+    - Define a new AEAD-like interface, i.e. `AEAD2`, whose `Seal` method does not accept an IV and returns a byte slice with iv+ciphertext+tag.
+    - Construct the AES-GCM TLS cipher using the new method.
+    - Don't generate an IV when the cipher implements the AEAD2 interface.
+    - Use the AEAD2 interface if available when encrypting.
+    - Test and validate the new path.
+
+2. Use OpenSSL AES-GCM in non-TLS mode, which means Go constructs the IV and OpenSSL honors it, but then we miss some required security checks that we would have to implement ourselves, namely:
+  - Enforce strictly monotonically increasing explicit nonces.
+  - Enforce explicit nonce values to be less than 2^64 - 1.
+  - Tets the new checks.
+
+#### Resolution
+
+We will implement option 2.
+
+Reasoning:
+- The additional security checks are easy to implement.
+- The changes are contained in `go-crypto-openssl`.
+- Option 1 would require non-trivial patching of the Go TLS stack, which we would need to keep up-to-date when `crypto/tls` changes.
+- Option 1 has higher chances to introduce a behavior chance or a security bug.
+
+Drawbacks:
+- OpenSSL might implement additional IV checks in future versions when running in TLS mode, and we will not benefit from them. We will have to keep an eye out for changes in this area.
+- We will implement security checks that go outside the ideal scope of `go-crypto-openssl`, which is just to act at as a thin layer between Go and OpenSSL APIs without real security knowledge.
```

Comment:
I think simplifying makes it clearer on first read that the `which` is about the scope, not the security checks we're implementing:

```suggestion
- Adding security checks goes outside the ideal scope of `go-crypto-openssl`, which is just to act at as a thin layer between Go and OpenSSL APIs without real security knowledge.
```